src/main/java/org/opensearch/security/http/OnBehalfOfAuthenticator.java

```diff
@@ -0,0 +1,253 @@
+import org.opensearch.security.ssl.util.ExceptionUtils;
+import org.opensearch.security.user.AuthCredentials;
+import org.opensearch.security.util.keyUtil;
+
+import static org.opensearch.security.OpenSearchSecurityPlugin.LEGACY_OPENDISTRO_PREFIX;
+import static org.opensearch.security.OpenSearchSecurityPlugin.PLUGINS_PREFIX;
+
+public class OnBehalfOfAuthenticator implements HTTPAuthenticator {
+
+    private static final String REGEX_PATH_PREFIX = "/(" + LEGACY_OPENDISTRO_PREFIX + "|" + PLUGINS_PREFIX + ")/" + "(.*)";
+    private static final Pattern PATTERN_PATH_PREFIX = Pattern.compile(REGEX_PATH_PREFIX);
+    private static final String ON_BEHALF_OF_SUFFIX = "api/user/onbehalfof";
+    private static final String ACCOUNT_SUFFIX = "api/account";
+
+    protected final Logger log = LogManager.getLogger(this.getClass());
+
+    private static final Pattern BEARER = Pattern.compile("^\\s*Bearer\\s.*", Pattern.CASE_INSENSITIVE);
+    private static final String BEARER_PREFIX = "bearer ";
+    private static final String TOKEN_TYPE_CLAIM = "typ";
+    private static final String TOKEN_TYPE = "obo";
+
+    private final JwtParser jwtParser;
+    private final String encryptionKey;
+    private final Boolean oboEnabled;
+
+    public OnBehalfOfAuthenticator(Settings settings) {
+        String oboEnabledSetting = settings.get("enabled");
+        oboEnabled = oboEnabledSetting == null ? Boolean.TRUE : Boolean.valueOf(oboEnabledSetting);
+        encryptionKey = settings.get("encryption_key");
+        jwtParser = initParser(settings.get("signing_key"));
+    }
+
+    private JwtParser initParser(final String signingKey) {
+        JwtParser _jwtParser = keyUtil.keyAlgorithmCheck(signingKey, log);
+
+        if (_jwtParser == null) {
+            throw new RuntimeException("Unable to find on behalf of authenticator signing key");
+        }
+
+        return _jwtParser;
+    }
+
+    private List<String> extractSecurityRolesFromClaims(Claims claims) {
+        Object rolesObject = ObjectUtils.firstNonNull(claims.get("er"), claims.get("dr"));
+        List<String> roles;
+
+        if (rolesObject == null) {
+            log.warn("This is a malformed On-behalf-of Token");
+            roles = List.of();
+        } else {
+            final String rolesClaim = rolesObject.toString();
+
+            // Extracting roles based on the compatbility mode
+            String decryptedRoles = rolesClaim;
+            if (rolesObject == claims.get("er")) {
+                decryptedRoles = EncryptionDecryptionUtil.decrypt(encryptionKey, rolesClaim);
+            }
+            roles = Arrays.stream(decryptedRoles.split(",")).map(String::trim).collect(Collectors.toList());
+        }
+
+        return roles;
+    }
+
+    private String[] extractBackendRolesFromClaims(Claims claims) {
+        // Object backendRolesObject = ObjectUtils.firstNonNull(claims.get("ebr"), claims.get("dbr"));
+        if (!claims.containsKey("dbr")) {
```

Comment:
wdyt about having this be simply `br` for backend_roles? I know in the past we discussed sending backend roles as encrypted/decrypted depending on if the extension was in backward compatibility mode, but on second thought would encrypted backend roles need to be sent at all?

IMO if the setting for backward comptability mode (or `include_backend_roles` or w/e we call it) is true then we can include `br` as a claim in the token and they would be in plain text. Roles, on the other hand, would need to be sent in all cases.